src/rest/auth.ts

```diff
@@ -0,0 +1,178 @@
+import Fastify, { FastifyInstance } from "fastify";
+import { ErrorCodes, sendError } from "../common/error";
+import { psqlClient } from "../common/database";
+import { compare, genSalt, hash } from "bcrypt";
+import { User } from "../common/types";
+import { addToken, checkCredentials } from "../common/auth";
+import { validateCaptcha } from "../common/captcha";
+import { moderateMessage } from "../common/moderate";
+import { salt } from "../common/constants";
+import { generateID } from "../common/ids";
+import { randomBytes } from "crypto";
+import { sendEmail } from "../common/email";
+
+type LoginBody = {
+    email: string;
+    password: string;
+};
+type SignupBody = {
+    username: string;
+    email: string;
+    password: string;
+    turnstileKey: string;
+};
+
+async function plugin(fst: FastifyInstance, opts) {
+    fst.post(
+        "/login",
+        {
+            schema: {
+                body: {
+                    type: "object",
+                    required: ["email", "password"],
+                    properties: {
+                        email: { type: "string", format: "email" },
+                        password: { type: "string" }
+                    }
+                }
+            }
+        },
+        async (req, res) => {
+            const body = req.body as LoginBody;
+            const auth = await checkCredentials(body.email, body.password);
+            if (!auth.success) {
+                switch (auth.error) {
+                    case "accountNotActivated":
+                        return sendError(
+                            res,
+                            "rest",
+                            ErrorCodes.AuthError,
+                            "Check your email to activate your account"
+                        );
+                    case "invalidCredentials":
+                        return sendError(
+                            res,
+                            "rest",
+                            ErrorCodes.AuthError,
+                            "Invalid username or password"
+                        );
+                }
+            }
+            const token = await addToken(auth.id);
+            return {
+                token
+            };
+        }
+    );
+
+    fst.post(
+        "/signup",
+        {
+            schema: {
+                body: {
+                    type: "object",
+                    required: ["username", "email", "password", "turnstileKey"],
+                    properties: {
+                        username: { type: "string", minLength: 2, maxLength: 30 },
+                        email: { type: "string", format: "email" },
+                        password: { type: "string" },
+                        turnstileKey: { type: "string" }
```

Comment:
style: trailing comma on multiline objects for better diffs
```suggestion
                        turnstileKey: { type: "string" },
```